Compute required beatmap attributes in DifficultyCalculators

smoogipoo opened the pull request:
The primary goal of this set of changes is to output beatmap attributes from the difficulty calculators for the performance calculator to consume. The reason for this is two ways:
1. The diffcalc server will be able to overwrite all attributes currently in the DB.
2. An osu!lazer-based performance calculator at some point in the future will not need to parse the beatmap, and instead be able to read attributes from the DB as is currently done in osu-performance.

smoogipoo committed:
Make performance calculators reference score for mods

smoogipoo committed:
Make difficulty/performance calculators provide object output data

smoogipoo committed:
Move osu! difficulty attributes to OsuDifficultyCalculator

smoogipoo committed:
Move osu!taiko difficulty attributes to TaikoDifficultyCalculator

smoogipoo committed:
Move osu!mania difficulty attributes to ManiaDifficultyCalculator

smoogipoo committed:
Remove now unneeded Beatmap field from performance calculators

smoogipoo committed:
Allow performance calculators to be constructed with difficulty attributes

smoogipoo committed:
Fix hitwindow naming

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into diffcalc-server-attribs

Tom94 commented:
I would like to comment on your second point: I think it'd actually be nice to have the performance calculator calculate difficulties on-the-fly as they are required (with caching). This would have the nice benefit of allowing us to use more complex difficulty metrics for performance computation than can be stored in the DB, such as per-object strain. I have a few idea on how such data could be useful without requiring scores to have per-hitobject data.

smoogipoo commented:
Might be useful indeed. I'm fine with removing my change if that's the path we should go down (perf calc always computing dificulty).

smoogipoo committed:
Give PerformanceCalculator a beatmap again

Following @Tom94's comment in https://github.com/ppy/osu/pull/2719, we'll probably want to compute difficulties on-the-fly in the future so more complex parameters can be utilised than the database can store provide.

peppy commented on osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs:
```diff
@@ -32,10 +32,6 @@ protected virtual void ApplyMods(Mod[] mods)
-        }
-
-        public abstract double Calculate(Dictionary<string, double> categoryDifficulty = null);
+        public abstract double Calculate(Dictionary<string, object> categoryDifficulty = null);
```
This could do with some documentation (what is the nullable dictionary parameter?)

smoogipoo committed:
Merge remote-tracking branch 'Joehuu/master' into diffcalc-server-attribs

# Conflicts:
#	osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs

smoogipoo committed:
Cleanups

smoogipoo committed:
Add xmldocs

peppy committed:
Merge branch 'master' into diffcalc-server-attribs

smoogipoo commented:
Closing for re-implementation of difficulty calculators.
